File: C/shop.c
// Karolina Szafran - Belzowska, November 2020
// Project Multi Paradigm Programming, Shop Assignment

// call libraries in C
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

// to compile and print shop.c in Command Prompt I have to type the following:
// gcc -o shop shop.c
// shop.exe


// A struct (keyword) is used to create a structure and group data together.
// A structure creates a data type that can be used to group items of possibly different types into a singe type.
// https://www.geeksforgeeks.org/structures-c/ 15/11/2020

struct Product  // Product Struct for product informations
{
	char* name; // name refers to a pointer, so when dereferenced, name is a char. In this case a pointer initialized to a string located in read-only memory.
	double price; //  Double = decimal number
};

struct ProductStock // ProductStock Struct for stock informations
{
	struct Product product;
	int quantity;
};

struct Shop // Shop Struct for shop infomations
{
	double cash;
	struct ProductStock stock[20];
	int index;
};

struct Customer // Customer Struct for customer informations
{
	char* name;
	double budget;
	struct ProductStock shoppingList[10]; // Nested struct - An Array of 10 elements.
	int index;
};

// Void when used as a function return type, the void keyword specifies that the function does not return a value.
// Void does not provide a result value to its caller. Those functions are called for their side effects, such as performing some task or writing to their output parameters.
// https://docs.microsoft.com/en-us/cpp/cpp/void-cpp?view=msvc-160, 15/11/2020
// https://en.wikipedia.org/wiki/Void_type, 15/11/2020


// ============================================================================================================================================================================

// printProduct() function will be used to print product informations
void printProduct(struct Product p)
{
	printf("Product's name: %s \nProduct's price: %.2f\n", p.name, p.price); // Print the product name and its price.
	printf("\n");
}

struct Shop createAndStockShop()
{
	
    FILE * fp;
    char * line = NULL;
    size_t len = 0;
    ssize_t read;
	int i = 0;

    fp = fopen("stock1.csv", "r");
    if (fp == NULL)
        exit(EXIT_FAILURE);

	getline(&line, &len, fp);
	// read cash in 
	double cashInshop = atof(line);
	struct Shop shop = { cashInshop };


	while ((read = getline(&line, &len, fp)) != -1)
	{
            char *n = strtok(line, ",");
            char *p = strtok(NULL, ",");
            char *q = strtok(NULL, ",");
            // convert to datatypes
            double price = atof(p);
            int quantity = atoi(q);
            char *name = malloc(sizeof(char)*50);
            strcpy(name, n);

			
            struct Product product = {name, price};
            struct ProductStock stockItem = {product, quantity};
            shop.stock[shop.index++]=stockItem;
    }

	return shop;
}

void printShop(struct Shop s) // This will print shop details like cash and stock available
{
	printf("\n\n\n------------------------------------------------------------\n                  WELCOME TO THE 'C' SHOP \n------------------------------------------------------------\n ");
	printf("**********************************************************\n");

    printf("\n");
	printf("============================================================\n");
    printf("   Cash in the Shop: EUR %.2f \n", s.cash); // print how much cash Store has in cash. Taken from stock.csv file
    printf("============================================================\n");
	printf("   Credit/Debit card payments acceptable in this Shop\n"); // just to inform customers that card payments are avilable
	printf("============================================================\n");
	printf("\n");
	printf("\n");
    printf("\n*** STOCK AVAILABLE IN 'C' SHOP ***\n");
    printf("\n");

	for (int i = 0; i < s.index; i++) // 'for' loop displays the stock details to the output
	{
		
		printProduct(s.stock[i].product);
		printf("Quantity: %d\n", s.stock[i].quantity); // When you press 1st option in Menu selection, this quantity will be given as output.

		printf("-------------------------------------------------------\n");
	}
}

struct Customer customerOrder() // this struct will read order.csv file  
{
	
	FILE * fp;
	char * line = NULL;
	size_t len = 0;
	ssize_t read;
	int i = 0;

	fp = fopen("order.csv", "r"); // fopen reads the customer's order csv file
	
	if (fp == NULL) // error exception
		exit(EXIT_FAILURE);

	getline(&line, &len, fp); // read customer name and budget, so csv file
	
	char *n = strtok(line, ",");
    char *b = strtok(NULL, ",");
	
	char *name = malloc(sizeof(char)*50); //create a local token string variable
    strcpy(name, n);
	double budget = atoi(b);
	struct Customer c ={ name, budget };
	
	// printf("CUSTOMER NAME: %s \nBUDGET: %.2f\n", c.name, c.budget);
	
    while ((read = getline(&line, &len, fp)) != -1)
    {
            char *r = strtok(line, ",");
			char *q = strtok(NULL, ",");
			int quantity = atoi(q);
			char *productname = malloc(sizeof(char)*50);
			strcpy(productname, r);
			struct Product product = {productname};
			struct ProductStock orderItem = {product, quantity};
			c.shoppingList[c.index++]=orderItem;
        }

    return c;
}

//================================================================================================================================

// printCustomer() will be used to print customer's details and order
void printCustomer(struct Customer c) 
{
	printf("\n=======================================================");
	printf("\nCustomer's name: %s \nCustomer's budget: EUR %.2f\n", c.name, c.budget); // Prints customer's name and budget
    printf("=======================================================\n");
	printf("\n");

	for (int i = 0; i < c.index; i++) // Start loop through each of the items in the shopping list 
	{
		printProduct(c.shoppingList[i].product); // Print the product informations
		printf("The customer wants %d of above\n", c.shoppingList[i].quantity);
		printf("\n-------------------------------------------------------\n");
	}
}

// =====================================================================================================================================

// searching both and comparing and returning price
double findPrice(struct Shop s, struct Customer c)
{
	struct Product productcustomer;
	double sum =0;
	double shortamount;
	
    printf("\n");
    printf("\n=======================================================\n");
	printf("\n      ***  Total to pay by customer  ***  \n");
    printf("=======================================================\n");
    printf("\n");

	for (int i = 0; i < s.index; i++)
	{
		for (int j = 0; j < c.index; j++) // loop through customer order
		{
			struct Product product = s.stock[i].product;
			struct Product productcustomer = c.shoppingList[j].product;
			// compare shop and order
			if (strcmp(product.name, productcustomer.name) == 0)
			{ 
					printf("Product: %s\nPrice: EUR %.2f * Quantity %d\n", c.shoppingList[j].product.name, product.price, c.shoppingList[j].quantity);
				    // check if in stock in store
					printf("In stock: %s, availability: %d\n", c.shoppingList[j].product.name, s.stock[i].quantity);
					printf("-------------------------------------------------------\n");
                    printf("\n");

					if(s.stock[i].quantity < c.shoppingList[j].quantity) // check if Shop has the item quantity
					{
                        printf("Sorry, the Shop has only %d \n \n", s.stock[i].quantity); // if not in stock
                        printf("-----------------------------------------------------\n");
                    }
					else{
				     
					sum = product.price * c.shoppingList[j].quantity + sum; // adding up the bill
					s.stock[i].quantity = s.stock[i].quantity - c.shoppingList[j].quantity;
					
					//i++;		
					}
			}		
		}
	}
		if(c.budget< sum)
		{
			shortamount = (sum - c.budget);
			printf("-----------------------------------------------------------------------------------------\n");
			printf("We are sorry, alternative tender required or you might need more money to pay: EUR %.2f \n", shortamount );
			printf("-----------------------------------------------------------------------------------------\n");
			printf("=========================================================================================\n");
							
		}
		else{
			printf("\n");
			printf("Total to pay by customer: EUR %.2f\n", sum);
            printf("===================================\n");
            printf("\n\n");
            printf("=======================================================\n");
			
            printf("\n*** Cash and availability of products in the Shop ***  \n");
            printf("=======================================================\n");
            printf("\n");
			
			
			s.cash = s.cash + sum; // add customer's bill to the shop's float
			printf("Cash has been updated and now is: EUR %.2f\n", s.cash);
			printf("\n-------------------------------------------------------\n");
			for (int i = 0; i < s.index; i++)
			{
				printf("Product's name: %s\nProduct's quantity after shopping: %d \n",s.stock[i].product.name, s.stock[i].quantity );
                printf("\n");
            }

		}

	return -1; 
}

// ========================================================================================================================================
struct Customer orderlive()
{
	printf("\n");
	printf("           *** Welcome to the Live Shop ***              \n");
	printf("=======================================================\n");
	printf("What is your name? "); // live Shop will ask for your name
	char *name = malloc(sizeof(char)*50);
    scanf(" %s", name);

	printf("\n");
	printf("What is your budget? "); // live Shop will ask for budget
	double budget;
	scanf("%lf", &budget);
	
	printf("\n");
	printf("What would you like to buy? "); // live Shop will ask for product you would like to buy
	char *p = malloc(sizeof(char)*50);
	char temp;
	scanf("%c", &temp); 
	scanf("%[^\n]s", p); 
	  
	printf("\n");
	printf("How many would you like to buy? "); // live Shop will ask how many products ypu would like to buy
	int q;
	scanf(" %d", &q);
	
	printf("------------------------------------------------------");
	struct Customer nc ={name, budget};
	printf("\nCustomer's name: %s\nCustomer's budget: %.2f",nc.name,nc.budget );

	struct Product product = {p};
	struct ProductStock orderItem = {product, q};
	nc.shoppingList[nc.index++]=orderItem;

	printf("\nProduct's name: %s\nquantity: %d", nc.shoppingList[0].product.name,nc.shoppingList[0].quantity );

	
   return nc;
}

// ===========================================================================================================================================
// int main(void) is the beginning of a function definition, (void) is the list of parameters that main is expecting from its caller. 
// In most cases, its caller is the operating system.

int main(void)
{
	
    struct Shop shop = createAndStockShop();
    printShop(shop);

    struct Customer c = customerOrder();
    printCustomer(c);
	
	findPrice(shop, c);


	struct Customer nc = orderlive();
	findPrice(shop, nc );
	

    return 0;
}

// https://stackoverflow.com/questions/12225171/difference-between-int-main-and-int-mainvoid, 18/11/2020

File: Python/shop.py
# Karolina Szafran - Belzowska, November 2020
# Project Multi Paradigm Programming, Shop Assignment

# call libraries in Python
from dataclasses import dataclass, field
from typing import List
import csv

# to print shop.py in Command Prompt I have to type the following:
# python shop.py

@dataclass
class Product:
    name: str
    price: float = 0.0

@dataclass 
class ProductStock:
    product: Product
    quantity: int

@dataclass 
class Shop:
    cash: float = 0.0
    stock: List[ProductStock] = field(default_factory=list)

@dataclass
class Customer:
    name: str = ""
    budget: float = 0.0
    shopping_list: List[ProductStock] = field(default_factory=list)

def create_and_stock_shop():
    s = Shop()
    with open('stock1.csv') as csv_file:
        csv_reader = csv.reader(csv_file, delimiter=',')
        first_row = next(csv_reader)
        s.cash = float(first_row[0])
        for row in csv_reader:
            p = Product(row[0], float(row[1]))
            ps = ProductStock(p, float(row[2]))
            s.stock.append(ps)
            # print(ps)
    return s
    
def read_customer(file_path):
    with open(file_path) as csv_file:
        csv_reader = csv.reader(csv_file, delimiter=',')
        first_row = next(csv_reader)
        c = Customer(first_row[0], float(first_row[1]))
        for row in csv_reader:
            name = row[0]
            quantity = float(row[1])
            p = Product(name)
            ps = ProductStock(p, quantity)
            c.shopping_list.append(ps)
        return c 
        
def print_product(p):
    print(f'\nPRODUCT NAME: {p.name} \nPRODUCT PRICE: {p.price}')

def print_product(p):
    print(f'\nProduct name: {p.name} \nProduct price: {p.price}')


def print_customer(c, s):
    check_stock(c, s)
    calculate_costs(c, s)
    total_order = 0

    print(f'======================================================')
    print(f'======================================================\n')
    print(f'Customer name: {c.name} \nCustomer budget: {c.budget}')
    print(f'------------------------------\n')

    for item in c.shopping_list:
        print_product(item.product)
        print(f'{c.name} wants {item.quantity:.0f} of above product')
        cost = item.quantity * item.product.price
        total_order += cost
        print(f'The cost to {c.name} will be EUR{cost:.2f}\n')

    if total_order <= c.budget:
        s.cash += total_order
        print("\n")
        print(f'Summary of the customers shopping basket:')
        print(f'==========================================================================')
        print(f'Customer name is {c.name} and she/he has EUR{c.budget:.2f} for the shopping.\nThe total price of the order for the customer is EUR{total_order:.2f}. \nTrasnaction complete. The customer now has {c.budget-total_order:.2f} remaining in the budget. \nCash in the Shop is now {s.cash}\n')
        for item in c.shopping_list:
            for prod in s.stock:
                if item.product.name == prod.product.name:
                    prod.quantity = prod.quantity - item.quantity
    else:
        print(f'Customer name is {c.name} and she/he has EUR{c.budget:.2f} for the shopping.\nThe total price of the order for {c.name} is €{total_order:.2f}. {c.name} has not enough money to complete the transaction. {s.cash}\n')

def check_stock(c, s):
    for item in c.shopping_list:
        for prod in s.stock:
                if item.product.name == prod.product.name and item.quantity <= prod.quantity:
                    print(item, item.quantity, prod.quantity)
                    print("OK")
                
                elif item.product.name == prod.product.name and item.quantity > prod.quantity:
                    print(f"We do not have enough stock of {item.product.name}, please re-select products to continue with your purchase.")
                    main()

def calculate_costs(c, s):
    for shop_item in s.stock:
        for list_item in c.shopping_list:
            if (list_item.product.name == shop_item.product.name):
                list_item.product.price = shop_item.product.price

def live_mode():
    print("=================================================\n")
    print("  ***Welcome to the Live Shop***   \n")
    print("=================================================\n")
    cust_name = input("What is your name? ") # live shop will  ask for your name
    budget= float(input(f"What is your budget {cust_name}?: "))
    c = Customer(cust_name, budget)
    print("Today we can offer you products which are listed below:")
    print_shop(s)
    shopping_list=[]
    additional_items = "Y"
    while (additional_items == "Y"):
        name = input("What would you like to buy?: ")
        quantity = int(input("How many would you like to buy?: "))
        p = Product(name)
        ps = ProductStock(p, quantity)
        c.shopping_list.append(ps)
        additional_items = input("Would you like to order additional items? Y/N \n")

    return c

def print_shop(s):
    print(f'-------------------------------------------------------\n           Welcome to the Python Shop \n-------------------------------------------------------')
    print(f'*******************************************************')
    print(f'\n')
    print(f'=======================================================\n')
    print(f'   Cash in the Shop: EUR {s.cash}\n')
    print(f'')
    print(f'   Credit/Debit card payments acceptable in this Shop') # just to inform customers that card payments are avilable
    print(f'=======================================================\n')
    print(f'*** STOCK AVAILABLE IN PYTHON SHOP *** ')
    print(f'---------------------------------------')

    for item in s.stock:
        print_product(item.product)
        print(f'The Shop has {item.quantity} of the above')

def main():
    live_mode()
   
    
    print("======================================================")
    print("======================================================")
    print("      Thank you for shopping in Python Shop")
    print("======================================================")


s = create_and_stock_shop()
print_shop(s)

c = read_customer("order.csv")
print_customer(c, s)

if __name__ == "__main__":
    
    main()